File: ext/go/src/aoc.com/tcollier/executor.go
package tcollier

import (
	"fmt"
	"time"
)

type Input interface{}

type solverFunc func(Input) string

type timingInfo struct {
	Iterations uint32
	Duration   int64
}

func continueTiming(iterations uint32, duration int64) bool {
	if iterations < 100 {
		return duration < 30000000
	} else {
		return duration < 100000
	}
}
func timeFn(input Input, fn solverFunc) timingInfo {
	var i uint32
	var runningTime int64
	for continueTiming(i, runningTime) {
		var startTime = time.Now().UnixNano()
		fn(input)
		runningTime += (time.Now().UnixNano() - startTime) / 1000
		i++
	}
	return timingInfo{Iterations: i, Duration: runningTime}
}

// Executor runs the solver code
func Executor(input Input, part1 solverFunc, part2 solverFunc, args []string) {
	if len(args) >= 1 && args[len(args)-1] == "--time" {
		var part1Time = timeFn(input, part1)
		var part2Time = timeFn(input, part2)
		fmt.Printf("{\"part1\":{\"iterations\":%d,\"duration\":%d},\"part2\":{\"iterations\":%d,\"duration\":%d}}\n", part1Time.Iterations, part1Time.Duration, part2Time.Iterations, part2Time.Duration)
	} else {
		fmt.Println(part1(input))
		fmt.Println(part2(input))
	}
}

File: ext/python/executor.py
import json

from datetime import datetime
from typing import Callable, Dict, List, Union

Input = Union[List[str], List[int]]
SolverCallable = Callable[[Input], str]


class Executor:
    def __init__(self, data: Input, part1_fn: SolverCallable, part2_fn: SolverCallable):
        self.data = data
        self.part1_fn = part1_fn
        self.part2_fn = part2_fn

    def __call__(self, argv: List[str]):
        if argv[-1] == "--time":
            self._time()
        else:
            self._solve()

    def _solve(self):
        print(self.part1_fn([v for v in self.data]))
        print(self.part2_fn([v for v in self.data]))

    def _time(self):
        print(
            json.dumps(
                {
                    "part1": self._time_fn(self.part1_fn),
                    "part2": self._time_fn(self.part2_fn),
                }
            )
        )

    def _time_fn(self, fn: SolverCallable) -> Dict:
        def duration_us(duration):
            return duration.seconds * 1000 * 1000 + duration.microseconds

        def continue_timing(iterations, duration):
            if iterations < 100:
                return duration < 30000000
            else:
                return duration < 100000

        i = 0
        running_time = 0
        while continue_timing(i, running_time):
            data = [v for v in self.data]
            start_time = datetime.now()
            fn(data)
            running_time += duration_us(datetime.now() - start_time)
            i += 1
        return {"duration": running_time, "iterations": i}
